### PR TITLE
feat(refactor): query_slurm function

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Refactor query_slurm and update_analysis_from_slurm_run_status functions

### Changed

- New utility function to return reformatted SLURM job ids
- New utility function to update analysis status from SLURM run
- Migrate and refactor query_slurm to slurm/api.py
- Extract method from update_analysis_from_slurm_run_status into CRUD
- Remove query_slurm from MockStore
- Add new test fixtures

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b refactor_query_slurm -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
Mere to add pytest-mock to requirements

File: trailblazer/server/api.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def update_analysis(analysis_id):
     """Update a specific analysis"""
     try:
         process = multiprocessing.Process(
-            target=store.update_run_status, kwargs={"analysis_id": analysis_id, "ssh": True}
+            target=store.update_run_status, kwargs={"analysis_id": analysis_id}
```
Comment (Vince-janv) [inline trailblazer/server/api.py:126]: Out of curiosity, what happens here? From the API we interact with the database via multiprocessing?

File: trailblazer/store/crud/update.py [hunk 2]
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
...
         ]
         self.commit()
+
+    def _update_analysis_status_from_slurm_jobs(self, analysis: Analysis) -> None:
+        """Update analysis status based on current SLURM jobs status."""
+        squeue_result: SqueueResult = get_squeue_result(
+            squeue_response=get_slurm_squeue_output(slurm_job_id_file=Path(analysis.config_path))
+        )
+        self.update_analysis_jobs_from_slurm_jobs(analysis=analysis, squeue_result=squeue_result)
+        LOG.info(f"Status in SLURM: {analysis.family} - {analysis.id}")
```
Comment (Vince-janv) [inline trailblazer/store/crud/update.py:64]: Is this really needed when we log the new status below?

File: trailblazer/store/api.py [hunk 3]
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
...
         elif analysis.workflow_manager == WorkflowManager.SLURM.value:
-            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id, ssh=ssh)
+            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id)
 
-    def update_analysis_from_slurm_run_status(self, analysis_id: int, ssh: bool = False) -> None:
-        """Query slurm for entries related to given analysis, and update the analysis in the database."""
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+    def update_analysis_from_slurm_run_status(self, analysis_id: int) -> None:
+        """Query slurm for entries related to given analysis, and update the analysis in the database.
+        Raises: Exception if unable to parse SLURM jobs
```
Comment (Vince-janv) [inline trailblazer/store/api.py:163]: We don't raise an error in here, we only seem to catch, log and then set the analysis to Error

File: tests/cli/test_cli_core.py [hunk 4]
```diff
@@ -146,11 +150,21 @@ def test_cancel_not_running(
...
     """Test cancelling an analysis, which is not running."""
     caplog.set_level("ERROR")
 
+    # GIVEN SLURM squeue output for an analysis
+    mocker.patch(
+        FUNC_GET_SLURM_SQUEUE_OUTPUT_PATH,
+        return_value=subprocess.check_output(
+            ["cat", slurm_squeue_output.get(failed_analysis_case_name)]
+        ).decode("utf-8"),
+    )
```
Comment (Vince-janv) [inline tests/cli/test_cli_core.py:166]: Neat ⭐

**Review by Vince-janv — APPROVED** [review]
Looks good! 👍

File: trailblazer/apps/slurm/api.py [hunk 5]
```diff
@@ -1,10 +1,46 @@
...
 from trailblazer.apps.slurm.utils import formatters
 from trailblazer.constants import FileFormat, SlurmJobStatus, TrailblazerStatus
 from trailblazer.exc import EmptySqueueError
-from trailblazer.io.controller import ReadStream
+from trailblazer.io.controller import ReadFile, ReadStream
+
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
+    job_ids: List[str] = slurm_job_id_file_content.get(next(iter(slurm_job_id_file_content)))
```
Comment (fevac) [inline trailblazer/apps/slurm/api.py:14]: I find this one liner difficult to read

File: trailblazer/apps/slurm/api.py [hunk 6]
```diff
@@ -1,10 +1,46 @@
...
 from trailblazer.apps.slurm.models import SqueueResult
 from trailblazer.apps.slurm.utils import formatters
 from trailblazer.constants import FileFormat, SlurmJobStatus, TrailblazerStatus
 from trailblazer.exc import EmptySqueueError
-from trailblazer.io.controller import ReadStream
+from trailblazer.io.controller import ReadFile, ReadStream
+
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
```
Comment (fevac) [inline trailblazer/apps/slurm/api.py:13]:
```suggestion
    """Return a string of comma separated SLURM job ids to be used as input for squeue."""
```

File: trailblazer/apps/slurm/api.py [hunk 7]
```diff
@@ -1,10 +1,46 @@
...
+        file_format=FileFormat.YAML, file_path=slurm_job_id_file
+    )
+    slurm_jobs: str = _get_squeue_jobs_flag_input(
+        slurm_job_id_file_content=slurm_job_id_file_content
+    )
+    return (
+        subprocess.check_output(
+            [
+                "ssh",
+                "hiseq.clinical@hasta.scilifelab.se",
```
Comment (fevac) [inline trailblazer/apps/slurm/api.py:30]: with these changes we are always doing ssh to hasta, is that what we want?

File: trailblazer/store/api.py [hunk 8]
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
         if analysis.workflow_manager == WorkflowManager.TOWER.value:
             self.update_tower_run_status(analysis_id=analysis_id)
         elif analysis.workflow_manager == WorkflowManager.SLURM.value:
-            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id, ssh=ssh)
+            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id)
```
Comment (fevac) [inline trailblazer/store/api.py:159]: do we want to have a controller for this? not sure, up to you, I'm starting to see controllers everywhere lately and I wonder 😬

**Review by fevac — COMMENTED** [review]
Nice! 🌟 I only got half way. I'll continue tomorrow

File: trailblazer/store/api.py [hunk 9]
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
...
 
-    def update_analysis_from_slurm_run_status(self, analysis_id: int, ssh: bool = False) -> None:
-        """Query slurm for entries related to given analysis, and update the analysis in the database."""
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+    def update_analysis_from_slurm_run_status(self, analysis_id: int) -> None:
```
Comment (fevac) [inline trailblazer/store/api.py:161]:
```suggestion
    def update_slurm_run_status(self, analysis_id: int) -> None:
```

File: trailblazer/store/api.py [hunk 10]
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
...
+        """Query slurm for entries related to given analysis, and update the analysis in the database.
```
Comment (fevac) [inline trailblazer/store/api.py:162]:
```suggestion
        """Query SLURM for entries related to given analysis, and update the analysis in the database (update jobs, progress, status and logged_at fields).
```

File: trailblazer/store/crud/update.py [hunk 11]
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
...
+        LOG.debug(squeue_result.jobs)
+        analysis.progress = squeue_result.jobs_status_distribution.get(
```
Comment (fevac) [inline trailblazer/store/crud/update.py:66]: type hint

File: trailblazer/store/crud/update.py [hunk 12]
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
...
+            SlurmJobStatus.COMPLETED, 0.0
+        )
+        analysis.status = get_current_analysis_status(
```
Comment (fevac) [inline trailblazer/store/crud/update.py:69]: type hint

File: trailblazer/store/crud/update.py [hunk 13]
```diff
@@ -43,3 +54,21 @@ def update_analysis_jobs_from_slurm_jobs(
...
+            jobs_status_distribution=squeue_result.jobs_status_distribution
+        )
+        LOG.info(f"Updated status {analysis.family} - {analysis.id}: {analysis.status} ")
+        analysis.logged_at = datetime.now()
```
Comment (fevac) [inline trailblazer/store/crud/update.py:73]: type hint

File: tests/cli/test_cli_core.py [hunk 14]
```diff
@@ -457,11 +481,20 @@ def test_scan(
...
     """Test scanning for analyses and updating analysis status."""
     caplog.set_level("INFO")
 
+    mocker.patch(
+        FUNC_GET_SLURM_SQUEUE_OUTPUT_PATH,
+        return_value=subprocess.check_output(
+            ["cat", slurm_squeue_output.get(ongoing_analysis_case_name)]
+        ).decode("utf-8"),
+    )
+
```
Comment (fevac) [inline tests/cli/test_cli_core.py:497]: I see this mocker repeated in your tests, is it possible to make a fixture of it? I'm not sure about this suggestion feel free to disregard

File: tests/conftest.py [hunk 15]
```diff
@@ -251,3 +256,82 @@ def fixture_tower_task() -> TowerTask:
...
+
+
+@pytest.fixture(scope="session")
+def slurm_squeue_output() -> Dict[str, Path]:
+    """Return SLURM squeue output for analysis started via SLURM."""
+    return {
+        "blazinginsect": Path(
+            "tests",
+            "fixtures",
+            "squeue",
```
Comment (fevac) [inline tests/conftest.py:268]: can we have a `fixture_directory` so we don't have to repeat this so many times?

**Review by fevac — CHANGES_REQUESTED** [review]
🌟

File: tests/cli/test_cli_core.py (hunk 4 above)
Comment (henrikstranneheim) [reply tests/cli/test_cli_core.py:166]: 👍 . Was important to move it from the Store

File: tests/cli/test_cli_core.py (hunk 14 above)
Comment (henrikstranneheim) [reply tests/cli/test_cli_core.py:498]: I think it is best to use it in test. The  slurm_squeue_output is already a fixture and needs dynamic input from the actual test.

File: trailblazer/store/crud/update.py (hunk 11 above)
Comment (henrikstranneheim) [reply trailblazer/store/crud/update.py:70]: We have decided to not type hint non-self attributes

File: trailblazer/store/crud/update.py (hunk 12 above)
Comment (henrikstranneheim) [reply trailblazer/store/crud/update.py:73]: Same here

File: tests/conftest.py (hunk 15 above)
Comment (henrikstranneheim) [reply tests/conftest.py]: Good suggestion - fixed

File: trailblazer/store/api.py (hunk 10 above)
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: I think this is too specific and easily gets out of sync. I will UC slurm 👍

File: trailblazer/store/api.py (hunk 3 above)
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: Removed

File: trailblazer/store/api.py (hunk 9 above)
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: But we do not update the slurm run status (we do not control it). We update the analysis based on the current slurm run status.

File: trailblazer/store/api.py (hunk 8 above)
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: 😄 Not for this

File: trailblazer/apps/slurm/api.py (hunk 7 above)
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: No, I misread the code and this would have broken it. Reverted 👍

File: trailblazer/server/api.py (hunk 1 above)
Comment (henrikstranneheim) [reply trailblazer/server/api.py]: Seems like

Comment (henrikstranneheim) [reply trailblazer/server/api.py]: Actually slurm on Hasta

File: tests/conftest.py [hunk 16]
```diff
@@ -251,3 +262,23 @@ def fixture_tower_task() -> TowerTask:
...
+        "daringpidgeon": Path(squeue_dir, f"daringpidgeon_{file_postfix}").as_posix(),
+        "escapedgoat": Path(squeue_dir, f"escapegoat_{file_postfix}").as_posix(),
+        "fancymole": Path(squeue_dir, f"fancymole_{file_postfix}").as_posix(),
+        "happycow": Path(squeue_dir, f"happycow_{file_postfix}").as_posix(),
+        "lateraligator": Path(squeue_dir, f"lateraligator_{file_postfix}").as_posix(),
+        "liberatedunicorn": Path(squeue_dir, f"liberatedunicorn_{file_postfix}").as_posix(),
+        "nicemice": Path(squeue_dir, f"nicemice_{file_postfix}").as_posix(),
+        "rarekitten": Path(squeue_dir, f"rarekitten_{file_postfix}").as_posix(),
+        "trueferret": Path(squeue_dir, f"trueferret_{file_postfix}").as_posix(),
+    }
```
Comment (ChrOertlin) [inline tests/conftest.py:284]: Perhaps a bit easier on the eyes.
```suggestion
@pytest.fixture(scope="session")
def slurm_squeue_output(squeue_dir: Path) -> Dict[str, Path]:
    """Return SLURM squeue output for analysis started via SLURM."""
    file_postfix: str = f"squeue{FileExtension.CSV}"
    
    job_names = [
        "blazinginsect", "crackpanda", "cuddlyhen", "daringpidgeon", "escapedgoat",
        "fancymole", "happycow", "lateraligator", "liberatedunicorn", "nicemice",
        "rarekitten", "trueferret"
    ]
    
    squeue_output = {}
    
    for job_name in job_names:
        squeue_output[job_name] = Path(squeue_dir, f"{job_name}_{file_postfix}").as_posix()
    
    return squeue_output
```

File: trailblazer/apps/slurm/api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: They are unpacking the dict. Removed for a more explicit approach

File: trailblazer/apps/slurm/api.py [hunk 17]
```diff
@@ -1,10 +1,50 @@
...
+        "%A,%j,%T,%l,%M,%S",
+    ]
+    if use_ssh:
+        squeue_commands = ["ssh", "hiseq.clinical@hasta.scilifelab.se"] + squeue_commands
+        return (
+            subprocess.check_output(
+                squeue_commands,
+                universal_newlines=True,
+            )
+            .decode("utf-8")
```
Comment (ChrOertlin) [inline trailblazer/apps/slurm/api.py:43]:
```suggestion
            .decode(UTF8CONSTANT)
```
seen this come around a couple of times as hard coded string in the code

File: trailblazer/apps/slurm/api.py [hunk 18]
```diff
@@ -1,10 +1,50 @@
...
+    squeue_commands: List[str] = [
+        "squeue",
+        "--jobs",
+        slurm_jobs,
+        "--states=all",
+        "--format",
+        "%A,%j,%T,%l,%M,%S",
+    ]
+    if use_ssh:
+        squeue_commands = ["ssh", "hiseq.clinical@hasta.scilifelab.se"] + squeue_commands
```
Comment (ChrOertlin) [inline trailblazer/apps/slurm/api.py:37]: Do we want a constant for the email address?

File: trailblazer/apps/slurm/api.py [hunk 19]
```diff
@@ -1,10 +1,46 @@
...
 from trailblazer.constants import FileFormat, SlurmJobStatus, TrailblazerStatus
 from trailblazer.exc import EmptySqueueError
-from trailblazer.io.controller import ReadStream
+from trailblazer.io.controller import ReadFile, ReadStream
+
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return SLURM squeue jobs flag formatted input."""
+    job_ids: List[str] = slurm_job_id_file_content.get(next(iter(slurm_job_id_file_content)))
+    return ",".join(map(str, job_ids))
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: No, not with updated approch

File: trailblazer/store/crud/update.py (hunk 2 above)
Comment (henrikstranneheim) [reply trailblazer/store/crud/update.py]: Changed to debug.

File: trailblazer/apps/slurm/api.py (hunk 5 above)
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: Modified it.

File: trailblazer/apps/slurm/api.py (hunk 6 above)
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: Added

**Review by ChrOertlin — COMMENTED** [review]
Looks good, some style things

**Comment by henrikstranneheim** [comment]
@fevac @Vince-janv Ready for another review!

File: trailblazer/store/api.py [hunk 20]
```diff
@@ -200,32 +156,15 @@ def update_run_status(self, analysis_id: int, ssh: bool = False) -> None:
...
-            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id, ssh=ssh)
+            self.update_analysis_from_slurm_run_status(analysis_id=analysis_id, use_ssh=use_ssh)
 
-    def update_analysis_from_slurm_run_status(self, analysis_id: int, ssh: bool = False) -> None:
+    def update_analysis_from_slurm_run_status(
+        self, analysis_id: int, use_ssh: bool = False
+    ) -> None:
         """Query slurm for entries related to given analysis, and update the analysis in the database."""
-        analysis: Analysis = self.get_analysis_with_id(analysis_id=analysis_id)
+        analysis: Optional[Analysis] = self.get_analysis_with_id(analysis_id=analysis_id)
```
Comment (ChrOertlin) [inline trailblazer/store/api.py:165]: would checking for 
```Python 
if not analysis:
```
help us define a more accurate error or is the try, except covering this?

File: trailblazer/apps/slurm/api.py [hunk 21]
```diff
@@ -1,10 +1,50 @@
...
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return a string of comma separated SLURM job ids to be used as input for squeue jobs flag."""
+    job_ids: List[str] = []
+    for slurm_job_ids in slurm_job_id_file_content.values():
+        [job_ids.append(str(job_id)) for job_id in slurm_job_ids]
+    return ",".join(job_ids)
+
+
+def get_slurm_squeue_output(slurm_job_id_file: Path, use_ssh: bool = False) -> str:
```
Comment (Vince-janv) [inline trailblazer/apps/slurm/api.py:20]: I'm a bit confused regarding the ssh option. What are we using in the automation? I'm not sure we need to support both of we only use one 🤔

**Review by Vince-janv — APPROVED** [review]
Looks good! Just curious about the ssh option

File: trailblazer/store/api.py (hunk 9 above)
Comment (fevac) [reply trailblazer/store/api.py]: you are right. I do find the current name confusing though, the `run_status` part is misleading to me. But I don't have good alternatives to offer, `update_analysis_from_slurm_jobs` ?
 I leave it up to you, feel free to resolve

File: trailblazer/apps/slurm/api.py [hunk 22]
```diff
@@ -1,10 +1,50 @@
...
 from trailblazer.exc import EmptySqueueError
-from trailblazer.io.controller import ReadStream
+from trailblazer.io.controller import ReadFile, ReadStream
+
+
+def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]]) -> str:
+    """Return a string of comma separated SLURM job ids to be used as input for squeue jobs flag."""
+    job_ids: List[str] = []
+    for slurm_job_ids in slurm_job_id_file_content.values():
+        [job_ids.append(str(job_id)) for job_id in slurm_job_ids]
```
Comment (diitaz93) [inline trailblazer/apps/slurm/api.py:16]:
```suggestion
    slurm_job_ids: List[List[str]] = slurm_job_id_file_content.values()
    job_ids: List[str] = [job_id for job_id in job_ids for job_ids in slurm_job_ids]
```

Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py:21]: Not a fan of nested for loops within a list comprehension. I think it is hard to understand. I would like to keep the current implementation. Ok?

File: trailblazer/store/api.py (hunk 9 above)
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: 👍

File: trailblazer/store/api.py (hunk 20 above)
Comment (henrikstranneheim) [reply trailblazer/store/api.py:163]: It should be handled by the try. Probably need more refactoring to raise more specific exceptions in called functions, but out of scope here.

File: trailblazer/apps/slurm/api.py (hunk 21 above)
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py:25]: FE:
```
process = multiprocessing.Process(target=store.update_ongoing_analyses, kwargs={"use_ssh": True})
       
process = multiprocessing.Process(
            target=store.update_run_status, kwargs={"analysis_id": analysis_id, "use_ssh": True}
```

CLI:
```
context.obj["trailblazer"].update_run_status(analysis_id=analysis_id)
 and several other places
```

So both with and without depending who is calling.
Automation on Hasta do not use ssh

File: trailblazer/apps/slurm/api.py (hunk 17 above)
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: Fixed

File: trailblazer/constants.py [hunk 23]
```diff
@@ -75,6 +75,12 @@ def ongoing_statuses(cls) -> Tuple:
         return cls.PENDING.value, cls.RUNNING.value, cls.COMPLETING.value
 
 
+class CharacterFormat(str, Enum):
+    """Define xharacter encoding/decoding to use."""
```
Comment (henrikstranneheim) [inline trailblazer/constants.py:79]:
```suggestion
    """Define character encoding/decoding to use."""
```

File: trailblazer/server/api.py [hunk 24]
```diff
@@ -113,22 +113,24 @@ def aggregate_jobs():
 @blueprint.route("/update-all")
 def update_analyses():
     """Update all ongoing analysis by querying SLURM"""
```
Comment (fevac) [inline trailblazer/server/api.py:115]:
```suggestion
    """Update all ongoing analysis by querying SLURM."""
```

**Review by fevac — APPROVED** [review]
Very well done! So much nicer ❤️  Looks good to me

File: tests/store/test_store_api.py [hunk 25]
```diff
@@ -1,13 +1,16 @@
...
 
 import pytest
 
 from tests.apps.tower.conftest import CaseName
 from tests.mocks.store_mock import MockStore
-from trailblazer.constants import TrailblazerStatus
+from trailblazer.constants import CharacterFormat, TrailblazerStatus
 from trailblazer.store.models import Analysis
 
+FUNC_GET_SLURM_SQUEUE_OUTPUT_PATH: str = "trailblazer.store.crud.update.get_slurm_squeue_output"
```
Comment (diitaz93) [inline tests/store/test_store_api.py:12]: Should this constant be defined only once in a constant file?

**Review by diitaz93 — COMMENTED** [review]
Looks good 👍

File: tests/store/test_store_api.py (hunk 25 above)
Comment (henrikstranneheim) [reply tests/store/test_store_api.py:12]: It is only for particular tests and I think it makes sense to define it where it is mocked as it has a very special use.

**Comment by henrikstranneheim** [comment]
### CLI ✅ 

```
θ60° [henrik.stranneheim@hasta:~] [S_base] 2m7s $ trailblazer scan
Status in Tower: richurchin - 278279 - 6231OHy5Yery61
Updated status richurchin - 278279: qc
Updated status mintprawn - 278267: running
Updated status notedocelot - 278263: running
Error updating analysis for: case - finerrhino : PermissionError
Error updating analysis for: case - immuneamoeba : PermissionError
Error updating analysis for: case - modestakita : PermissionError
Error updating analysis for: case - famouscorgi : PermissionError
Error updating analysis for: case - topskite : PermissionError
Error updating analysis for: case - firstviper : PermissionError
Error updating analysis for: case - equalthrush : PermissionError
Error updating analysis for: case - openduck : PermissionError
Error updating analysis for: case - busyturkey : PermissionError
Error updating analysis for: case - strongsponge : PermissionError
Error updating analysis for: case - helpedegret : PermissionError
Error updating analysis for: case - nativeloon : PermissionError
Error updating analysis for: case - supercougar : PermissionError
Error updating analysis for: case - socialcrab : PermissionError
Error updating analysis for: case - activelion : PermissionError
Error updating analysis for: case - amusedspider : PermissionError
Error updating analysis for: case - greatbear : PermissionError
Error updating analysis for: case - aceviper : PermissionError
Error updating analysis for: case - brightape : PermissionError
Error updating analysis for: case - wittyghoul : PermissionError
Error updating analysis for: case - safeshiner : PermissionError
Error updating analysis for: case - brightegret : PermissionError
Error updating analysis for: case - fitloon : PermissionError
Error updating analysis for: case - decentcalf : PermissionError
Error updating analysis for: case - knownburro : PermissionError
Error updating analysis for: case - smoothlocust : PermissionError
Error updating analysis for: case - cuteprawn : PermissionError
Error updating analysis for: case - frankpigeon : PermissionError
Error updating analysis for: case - politeowl : PermissionError
Error updating analysis for: case - likedmutt : PermissionError
Error updating analysis for: case - wholegull : PermissionError
Error updating analysis for: case - promptowl : PermissionError
Error updating analysis for: case - nicewhale : PermissionError
Error updating analysis for: case - activedodo : PermissionError
Error updating analysis for: case - guidedmole : PermissionError
Error updating analysis for: case - busyturtle : PermissionError
Error updating analysis for: case - caringdrum : PermissionError
Error updating analysis for: case - lovedcaiman : PermissionError
Error updating analysis for: case - awakemarten : PermissionError
Error updating analysis for: case - neededbass : PermissionError
Error updating analysis for: case - divinesawfly : PermissionError
Error updating analysis for: case - modestoriole : PermissionError
Error updating analysis for: case - onekoi : PermissionError
Error updating analysis for: case - immunefawn : PermissionError
Error updating analysis for: case - manycow : PermissionError
Error updating analysis for: case - heroicmule : PermissionError
Error updating analysis for: case - inorca : PermissionError
Updated status aceracer - 278157: completed
Error updating analysis for: case - protahr : PermissionError
Error updating analysis for: case - trustysquid : PermissionError
Error updating analysis for: case - solidhusky : PermissionError
Error updating analysis for: case - moralbeagle : PermissionError
Error updating analysis for: case - cracktiger : PermissionError
Error updating analysis for: case - daringrat : PermissionError
Error updating analysis for: case - nativepossum : PermissionError
Error updating analysis for: case - savingstag : FileNotFoundError
Error updating analysis for: case - HW2V2DRX2 : FileNotFoundError
Error updating analysis for: case - divinetomcat : PermissionError
Error updating analysis for: case - handyshrimp : FileNotFoundError
Error updating analysis for: case - setcolt : PermissionError
Error updating analysis for: case - proudfish : PermissionError
Error updating analysis for: case - funnyaphid : FileNotFoundError
Error updating analysis for: case - onestag : PermissionError
Error updating analysis for: case - newgopher : PermissionError
Error updating analysis for: case - helpedlab : MissingSqueueOutput
Error updating analysis for: case - goodbengal : MissingSqueueOutput
Error updating analysis for: case - grandaphid : PermissionError
All analyses updated!
```

### API ✅ 
Screenshot 2023-08-31 at 14 35 34

Screenshot 2023-08-31 at 14 38 18

**Comment by henrikstranneheim** [comment]
Deployed in production:
```
Log deploy... done.
trailblazer, version 17.0.7
θ67° [henrik.stranneheim@hasta:~] [P_base] 1m33s $
```

**Pull request merged**